### DISCUSSION TITLE
Specifying string formats for OpenAPI generation

### DISCUSSION THREAD

**Discussion by ptrdom** [disneystreaming/smithy4s#1218] [Q&A]
There are multiple different string formats supported by OpenAPI - https://swagger.io/docs/specification/data-models/data-types/#format, I wonder - is there a way to specify in smithy4s what `format` value should be set for a string attribute in OpenAPI schema? If not, would adding an annotation for it be beneficial?

The use case I am thinking of is using [type refinements](https://disneystreaming.github.io/smithy4s/docs/codegen/customisation/refinements) to add `java.time.LocalDate` refinement for a string type, and have `format` for that string attribute set to `date` in generated OpenAPI schema.

**Comment by daddykotex** [comment]
Hi @ptrdom glad to see you here.

Before I answer, let's describe some things to make sure we're talking about the same stuff:
* smithy - is the IDL you use to describe your services
* smithy4s (this repo) - is the tool to generate code from smithy
* type refinements - a feature of smithy4s to tweak code generation
* alloy - upstream dependency with common smithy shapes and projects
* timestamp - there is a `Timestamp` type in smithy that renders to `format:date`

So right now, smithy4s generate scala code, but also open api spec, via: [alloy-openapi](https://github.com/disneystreaming/alloy/blob/1fe8d6df01f1d85b2c13213ec40d76b7d3ab6e0b/build.sc#L147). This package is build on top on [smithy-openapi](https://smithy.io/2.0/guides/converting-to-openapi.html?highlight=openapi) and so it's highly customizable.

So what you want to do is not directly supported by any of these but it can be implemented.

If you have, on the class path, an implementation of `Smithy2OpenApiExtension` along with the java service provider stuff so that `smithy-openapi` can load it (eg: `resources/META-INF/services/software.amazon.smithy.openapi.fromsmithy.Smithy2OpenApiExtension`, for examples [see what we do in alloy](https://github.com/disneystreaming/alloy/blob/1fe8d6df01f1d85b2c13213ec40d76b7d3ab6e0b/modules/openapi/resources/META-INF/services/software.amazon.smithy.openapi.fromsmithy.Smithy2OpenApiExtension#L1)), you can use it to do that I believe

So if I were to do something like you want, I'd probably to the following:
1. add a special trait that I can mark my string with
2. implement a `Smithy2OpenApiExtension` to process that and tweak the open api representation
3. make that available to the smithy4s class path so that it's picked up when generating the openapi (not entirely sure how I would do that just yet)

**Reply by ptrdom** [reply]
Aha, sounds doable, I am going try it out.

Thank you for a very detailed explanation and a quick reply!
